tcn-client-android/src/main/java/org/tcncoalition/tcnclient/crypto/Keys.kt

```diff
@@ -0,0 +1,168 @@
+package org.tcncoalition.tcnclient.crypto
+
+import cafe.cryptography.ed25519.Ed25519PrivateKey
+import cafe.cryptography.ed25519.Ed25519PublicKey
+import org.tcncoalition.tcnclient.TcnConstants.H_TCK_DOMAIN_SEPARATOR
+import org.tcncoalition.tcnclient.TcnConstants.H_TCN_DOMAIN_SEPARATOR
+import java.nio.ByteBuffer
+import java.nio.ByteOrder
+import java.security.MessageDigest
+import java.security.SecureRandom
+
+/** Authorizes publication of a report of potential exposure. */
+class ReportAuthorizationKey(internal val rak: Ed25519PrivateKey) {
+    internal val rvk: Ed25519PublicKey = this.rak.derivePublic()
+
+    /** Generates a new ReportAuthorizationKey. */
+    constructor(random: SecureRandom) : this(Ed25519PrivateKey.generate(random))
+
+    companion object Reader {
```

Comment:
If the idea was to use as a common object for serialization we could do something like this:
```
interface Reader<T> {
    fun fromByteArray(bytes: ByteArray): T
}
```

And then all classes implements in companion objects like this:
```
  companion object: Reader<TemporaryContactKey> {
        /** Reads a [TemporaryContactKey] from [bytes]. */
        override fun fromByteArray(bytes: ByteArray): TemporaryContactKey 
.....
```

Reply:
Another option is to override `invoke` function:
```
operator fun invoke(bytes: ByteArray) : ReportAuthorizationKey {
    val buf = ByteBuffer.wrap(bytes)
    val rak = Ed25519PrivateKey.fromByteArray(read32(buf))
    return ReportAuthorizationKey(rak)
}
```
And use it in code like this
```
ReportAuthorizationKey(rakBuf1).toByteArray()
```

Reply:
> Why do you have `Reader` name for the companion object? We could omit it would work fine.

I picked the `Reader` name because the choice was irrelevant to Kotlin, and if this were to be called at some point from Java then `ReportAuthorizationKey.Reader.fromByteArray` was nicer than `ReportAuthorizationKey.Companion.fromByteArray`.

> If the idea was to use as a common object for serialization

I originally attempted to follow the reference implementation in having all serialization defined in a single place (the same way I have `ReportAuthorizationKey.createReport()` defined in `Report.kt`), but I ran into the static method problem you mention.

That being said, the `interface Reader<T>` is more what I was after. In particular, each companion object could just implement `fromByteBuffer()`, and then the interface could provide a default `fromByteArray()` implementation. I'll rework to do this.

> Another option is to override `invoke` function:

I'm not sure this works; doesn't `invoke` require an already-instantiated object?

Reply:
>I picked the Reader name because the choice was irrelevant to Kotlin, and if this were to be called at some point from Java then ReportAuthorizationKey.Reader.fromByteArray was nicer than ReportAuthorizationKey.Companion.fromByteArray.

As I mentioned, if you don't provide the name it works as a static method in Kotlin if you want to call a method. Meaning instead of `ReportAuthorizationKey.Companion.fromByteArray` you call just `ReportAuthorizationKey.fromByteArray` and to call the same from java you just add @JvmStatic annotation for a method.

>That being said, the interface Reader<T> is more what I was after. In particular, each companion object could just implement fromByteBuffer(), and then the interface could provide a default fromByteArray() implementation. I'll rework to do this.

I'm not sure that I'm following this. I tried different combinations but I couldn't find how to make it more or less robust.

>I'm not sure this works; doesn't invoke require an already-instantiated object?
It will work but might create a confusion as it looks like an object being initiated but in fact it is an `invoke` operator which is `()`. This is similar to Swift's `init` method i guess.